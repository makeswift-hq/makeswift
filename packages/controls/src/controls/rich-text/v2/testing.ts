import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
