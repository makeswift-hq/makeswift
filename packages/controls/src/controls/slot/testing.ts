import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
