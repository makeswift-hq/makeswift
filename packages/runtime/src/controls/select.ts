export const SelectControlType = 'makeswift::controls::select'

type SelectControlOption<T extends string> = { value: T; label: string }

type SelectControlConfig<T extends string = string> = {
  label?: string
  labelOrientation?: 'horizontal' | 'vertical'
  options: SelectControlOption<T>[]
  defaultValue?: T
}

export type SelectControlDefinition<C extends SelectControlConfig = SelectControlConfig> = {
  type: typeof SelectControlType
  config: C
}

export type SelectControlDefinitionOption<T extends SelectControlDefinition> =
  T['config'] extends SelectControlConfig<infer U> ? U : never

export type SelectControlData<T extends SelectControlDefinition = SelectControlDefinition> =
  SelectControlDefinitionOption<T> | null

export function Select<T extends string, C extends SelectControlConfig<T>>(
  config: C & { options: SelectControlOption<T>[] },
): SelectControlDefinition<C> {
  return { type: SelectControlType, config }
}
