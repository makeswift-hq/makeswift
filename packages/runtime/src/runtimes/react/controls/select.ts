import { type DataType, type ResolvedValueType } from '@makeswift/controls'

import { SelectDefinition } from '../../../controls'

export function useSelectControlValue(
  data: DataType<SelectDefinition> | undefined,
  definition: SelectDefinition,
): ResolvedValueType<SelectDefinition> {
  return definition.fromData(data) ?? definition.config.defaultValue
}
