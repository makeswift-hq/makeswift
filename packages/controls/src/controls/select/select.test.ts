import { testDefinition } from '../../testing/test-definition'

import { Select } from './select'

const options = [
  { value: 'p', label: 'Paragraph' },
  { value: 'h1', label: 'Heading 1' },
] as const

describe('Select', () => {
  describe('constructor', () => {
    test('returns correct definition', () => {
      const blockType = Select({
        label: 'Block type',
        labelOrientation: 'horizontal',
        options,
        defaultValue: 'p',
      })

      expect(blockType).toMatchSnapshot()
    })

    test('enforces item value type', () => {
      Select({
        label: 'Block type',
        options: [
          {
            // @ts-expect-error
            value: 42,
            label: 'Paragraph',
          },
        ],
      })
    })

    test('disallows extraneous properties', () => {
      Select({
        options,
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    // FIXME
    // function assignTest(_def: SelectDefinition) {}
    // assignTest(Select({ options }))
    // assignTest(Select({ label: 'Block type', options }))
    // assignTest(Select({ options, defaultValue: 'p' }))
    // assignTest(Select({ options, defaultValue: 'p' as string }))
    // assignTest(Select({ options, defaultValue: undefined }))
    // assignTest(Select({ label: 'Block type', options, defaultValue: undefined }))
    // assignTest(Select({ label: undefined, options, defaultValue: undefined }))
  })

  const invalidValues = [null, 17, 'random', { swatchId: 42 }]

  testDefinition(
    Select({
      label: 'Block type',
      labelOrientation: 'horizontal',
      options,
    }),
    ['p', 'h1'],
    invalidValues,
  )
})
