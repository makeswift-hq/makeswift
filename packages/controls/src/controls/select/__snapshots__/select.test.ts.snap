// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select constructor returns correct definition 1`] = `
SelectDefinition {
  "config": {
    "defaultValue": "p",
    "label": "Block type",
    "labelOrientation": "horizontal",
    "options": [
      {
        "label": "Paragraph",
        "value": "p",
      },
      {
        "label": "Heading 1",
        "value": "h1",
      },
    ],
  },
}
`;

exports[`Select definition w/ config {"options":[{"value":"p","label":"Paragraph"},{"value":"h1","label":"Heading 1"}],"defaultValue":"h1"} toData returns v1 data for \`h1\` when definition version is 1 1`] = `"h1"`;

exports[`Select definition w/ config {"options":[{"value":"p","label":"Paragraph"},{"value":"h1","label":"Heading 1"}],"defaultValue":"h1"} toData returns v1 data for \`p\` when definition version is 1 1`] = `"p"`;

exports[`Select definition w/ config {"options":[{"value":"p","label":"Paragraph"},{"value":"h1","label":"Heading 1"}]} toData returns v1 data for \`h1\` when definition version is 1 1`] = `"h1"`;

exports[`Select definition w/ config {"options":[{"value":"p","label":"Paragraph"},{"value":"h1","label":"Heading 1"}]} toData returns v1 data for \`p\` when definition version is 1 1`] = `"p"`;
